lib/ansible/module_utils/dimensiondata.py

```diff
@@ -0,0 +1,393 @@
+                                       block.id, False)
+    pub_ips = expand_ip_block(pub_ip_block)
+    block_detailed = {'id': block.id, 'addresses': []}
+    for ip in pub_ips:
+        allocated = False
+        nat_match = filter(lambda x: x.external_ip == ip, nat_rules)
+        lb_match = filter(lambda x: x.ip == ip, balancers)
+        if len(nat_match) > 0 or len(lb_match) > 0:
+            allocated = True
+        else:
+            allocated = False
+        block_detailed['addresses'].append({'address': ip,
+                                            'allocated': allocated})
+    return block_detailed
+
+
+def list_balancers(module, lb_driver):
+    try:
+        return lb_driver.list_balancers()
+    except DimensionDataAPIException as e:
+        module.fail_json(msg="Failed to list Load Balancers: %s" % e.message)
+
+
+def get_blocks_with_unallocated(module, cp_driver, lb_driver, network_domain):
+    """
+    Gets ip blocks with one or more unallocated IPs.
+    ex:
+        {'unallocated_count': <total count of unallocated ips>,
+         'ip_blocks': [<list of expanded blocks with details
+                       (see get_block_allocation())>],
+         'unallocated_addresses': [<list of unallocated ip addresses>]
+        }
+    """
+    total_unallocated_ips = 0
+    all_blocks = list_public_ip_blocks(module, cp_driver, network_domain)
+    unalloc_blocks = []
+    unalloc_addresses = []
+    for block in all_blocks:
+        d_blocks = get_block_allocation(module, cp_driver, lb_driver,
+                                        network_domain, block)
+        i = 0
+        for addr in d_blocks['addresses']:
+            if addr['allocated'] is False:
+                if i == 0:
+                    unalloc_blocks.append(d_blocks)
+                unalloc_addresses.append(addr['address'])
+                total_unallocated_ips += 1
+            i += 1
+    return {'unallocated_count': total_unallocated_ips,
+            'ip_blocks': unalloc_blocks,
+            'unallocated_addresses': unalloc_addresses}
+
+
+def get_unallocated_public_ips(module, cp_driver, lb_driver, network_domain,
+                               reuse_free, count=0):
+    """
+    Get and/or provision unallocated public IPs
+    """
+    free_ips = []
+    if reuse_free is True:
+        blocks_with_unallocated = get_blocks_with_unallocated(module,
+                                                              cp_driver,
+                                                              lb_driver,
+                                                              network_domain)
+        free_ips = blocks_with_unallocated['unallocated_addresses']
+    if len(free_ips) < count:
+        num_needed = count - len(free_ips)
+        for i in range(num_needed):
+            block = cp_driver.ex_add_public_ip_block_to_network_domain(
+                network_domain)
+            b_dict = get_block_allocation(module, cp_driver, lb_driver,
```

Comment:
Just a style note... If you could use block_dict instead of b_dict that would be good.  We use the "b_" prefix as a convention to tell when we're dealing with byte strings as opposed to text strings.